File: grindx/judges/javascript/rotate-array.js
/**
 * Judge for Rotate Array — in-place mutation pattern.
 */

const path = require("path");
const { loadCases, reportAc, reportWa, loadSolution } = require("./_common");

function generateLargeCases() {
  const cases = [];
  const n = 100000;
  let seed = 42;
  const rand = () => {
    seed = (seed * 1103515245 + 12345) & 0x7fffffff;
    return (seed % 2000000001) - 1000000000;
  };

  // Large case 1: large k < n
  const nums1 = [];
  for (let i = 0; i < n; i++) nums1.push(rand());
  const k1 = Math.abs(rand()) % n;
  const ek1 = k1 % n;
  const expected1 = nums1.slice(n - ek1).concat(nums1.slice(0, n - ek1));
  cases.push({ input: [nums1, k1], expected: expected1, category: "tle" });

  // Large case 2: k > n
  const nums2 = [];
  for (let i = 0; i < n; i++) nums2.push(rand());
  const k2 = n + Math.abs(rand()) % n;
  const ek2 = k2 % n;
  const expected2 = nums2.slice(n - ek2).concat(nums2.slice(0, n - ek2));
  cases.push({ input: [nums2, k2], expected: expected2, category: "tle" });

  return cases;
}

function run(solutionPath) {
  const sol = loadSolution(path.resolve(solutionPath));
  const rotate = sol.rotate;
  if (typeof rotate !== "function") {
    process.stderr.write("Solution does not export a rotate function\n");
    process.exit(2);
  }

  const tc = loadCases("rotate-array");
  const basicCases = tc.cases;
  const largeCases = generateLargeCases();
  const allCases = basicCases.concat(largeCases);
  const total = allCases.length;

  for (let i = 0; i < allCases.length; i++) {
    const c = allCases[i];
    const [nums, k] = c.input;
    const numsCopy = [...nums];
    rotate(numsCopy, k);
    const expected = c.expected;

    if (JSON.stringify(numsCopy) !== JSON.stringify(expected)) {
      reportWa(i, c, numsCopy, total);
    }
  }
  reportAc(total);
}

run(process.argv[2]);

File: grindx/judges/python/rotate-array.py
"""Judge for Rotate Array — in-place mutation pattern."""

import importlib.util
import os
import random
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from _common import load_cases, report_ac, report_wa


def generate_large_cases() -> list[dict]:
    """Generate large array for TLE detection. Deterministic via seed."""
    random.seed(42)
    cases = []
    n = 100_000
    nums = [random.randint(-(10**9), 10**9) for _ in range(n)]
    k = random.randint(1, n - 1)
    # Rotate right by k: last k elements move to front
    expected = nums[-(k % n) :] + nums[: -(k % n)]
    cases.append(
        {
            "input": [nums, k],
            "expected": expected,
            "category": "tle",
        }
    )
    return cases


def run(solution_path: str) -> None:
    spec = importlib.util.spec_from_file_location("solution", solution_path)
    if spec is None or spec.loader is None:
        print(f"Could not load solution: {solution_path}", file=sys.stderr)
        sys.exit(2)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)

    tc = load_cases("rotate-array")
    basic_cases = tc["cases"]
    large_cases = generate_large_cases()
    all_cases = basic_cases + large_cases
    total = len(all_cases)

    for i, case in enumerate(all_cases):
        nums = list(case["input"][0])
        k = case["input"][1]

        mod.rotate(nums, k)

        if nums != case["expected"]:
            report_wa(i, case, nums, total)

    report_ac(total)


if __name__ == "__main__":
    run(sys.argv[1])
